Sources/SwiftProtobuf/SetPathDecoder.swift

```diff
@@ -0,0 +1,343 @@
+// Sources/SwiftProtobuf/SetPathDecoder.swift - Path decoder (Setter)
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Decoder which sets value of a field by its path.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+/// Describes errors can occure during decoding a proto by path.
+public enum PathDecodingError: Error {
+
+  /// Describes a mismatch in type of the fields.
+  ///
+  /// If a value of type A is applied to a path with type B.
+  /// this error will be thrown.
+  case typeMismatch
+}
+
+extension Message {
+  static func number(for field: String) -> Int? {
+    guard let type = Self.self as? _ProtoNameProviding.Type else {
+      return nil
+    }
+    return type._protobuf_nameMap.number(forJSONName: field)
+  }
+}
+
+struct SetPathDecoder<T: Message>: Decoder {
+
+  private let path: String
+  private let value: Any?
+  private var number: Int?
+
+  init(path: String, value: Any?) {
+    self.path = path
+    self.value = value
+    if let firstPathComponent {
+      self.number = T.number(for: firstPathComponent)
+    }
+  }
+
+  var firstPathComponent: String? {
+    return path
+          .components(separatedBy: ".")
+          .first
+  }
+
+  var nextPath: String {
+    return path
+          .components(separatedBy: ".")
+          .dropFirst()
+          .joined(separator: ".")
+  }
+
+  func _value<V>(as: V.Type) throws -> V {
+    guard let __value = self.value as? V else {
+      throw PathDecodingError.typeMismatch
+    }
+    return __value
+  }
+
+  mutating func handleConflictingOneOf() throws {}
+
+  mutating func nextFieldNumber() throws -> Int? {
+    defer { number = nil }
+    return number
+  }
+
+  mutating func decodeSingularFloatField(value: inout Float) throws {
+    value = try _value(as: Float.self)
+  }
+
+  mutating func decodeSingularFloatField(value: inout Float?) throws {
+    value = try _value(as: Float?.self)
+  }
+
+  mutating func decodeRepeatedFloatField(value: inout [Float]) throws {
+    value = try _value(as: [Float].self)
```

Comment:
The repeated handling appears to always being doing a full "set" vs. an "append", and some of the upstream docs seems to imply repeated fields should be appends for some operations.  Do we need a way to do either operation?

Reply:
It was a merge option in the file you sent and I didn't implement merge options as I said earlier. But this particular option could be implemented regardless of the other ones. I will go for it.

Reply:
I added `MergeOption` which has a case `replaceRepeatedFields` to determine it

Reply:
@thomasvl Did you check it? It's done